File: src/lib/storageService.ts
import { supabase } from './supabase';

export interface UploadResult {
  success: boolean;
  url?: string;
  path?: string;
  error?: string;
}

class StorageService {
  async uploadAvatar(userId: string, file: File): Promise<UploadResult> {
    try {
      const fileExt = file.name.split('.').pop();
      const fileName = `${userId}/${Date.now()}.${fileExt}`;

      const { error: uploadError } = await supabase.storage
        .from('avatars')
        .upload(fileName, file, {
          cacheControl: '3600',
          upsert: true,
        });

      if (uploadError) throw uploadError;

      const { data } = supabase.storage
        .from('avatars')
        .getPublicUrl(fileName);

      return {
        success: true,
        url: data.publicUrl,
        path: fileName,
      };
    } catch (error) {
      console.error('Avatar upload error:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Erreur lors de l\'upload',
      };
    }
  }

  async uploadPortfolioImage(userId: string, file: File): Promise<UploadResult> {
    try {
      const fileExt = file.name.split('.').pop();
      const fileName = `${userId}/${Date.now()}.${fileExt}`;

      const { error: uploadError } = await supabase.storage
        .from('portfolios')
        .upload(fileName, file, {
          cacheControl: '3600',
          upsert: false,
        });

      if (uploadError) throw uploadError;

      const { data } = supabase.storage
        .from('portfolios')
        .getPublicUrl(fileName);

      return {
        success: true,
        url: data.publicUrl,
        path: fileName,
      };
    } catch (error) {
      console.error('Portfolio upload error:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Erreur lors de l\'upload',
      };
    }
  }

  async uploadDocument(userId: string, file: File): Promise<UploadResult> {
    try {
      const fileExt = file.name.split('.').pop();
      const fileName = `${userId}/${Date.now()}.${fileExt}`;

      const { error: uploadError } = await supabase.storage
        .from('documents')
        .upload(fileName, file, {
          cacheControl: '3600',
          upsert: false,
        });

      if (uploadError) throw uploadError;

      const { data } = supabase.storage
        .from('documents')
        .getPublicUrl(fileName);

      return {
        success: true,
        url: data.publicUrl,
        path: fileName,
      };
    } catch (error) {
      console.error('Document upload error:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Erreur lors de l\'upload',
      };
    }
  }

  async uploadProjectPhoto(contractId: string, file: File): Promise<UploadResult> {
    try {
      const fileExt = file.name.split('.').pop();
      const fileName = `${contractId}/${Date.now()}.${fileExt}`;

      const { error: uploadError } = await supabase.storage
        .from('project-photos')
        .upload(fileName, file, {
          cacheControl: '3600',
          upsert: false,
        });

      if (uploadError) throw uploadError;

      const { data } = supabase.storage
        .from('project-photos')
        .getPublicUrl(fileName);

      return {
        success: true,
        url: data.publicUrl,
        path: fileName,
      };
    } catch (error) {
      console.error('Project photo upload error:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Erreur lors de l\'upload',
      };
    }
  }

  async deleteFile(bucket: string, path: string): Promise<boolean> {
    try {
      const { error } = await supabase.storage
        .from(bucket)
        .remove([path]);

      if (error) throw error;
      return true;
    } catch (error) {
      console.error('File deletion error:', error);
      return false;
    }
  }

  async listFiles(bucket: string, folder: string): Promise<string[]> {
    try {
      const { data, error } = await supabase.storage
        .from(bucket)
        .list(folder, {
          limit: 100,
          offset: 0,
          sortBy: { column: 'created_at', order: 'desc' },
        });

      if (error) throw error;
      return data?.map(file => `${folder}/${file.name}`) || [];
    } catch (error) {
      console.error('List files error:', error);
      return [];
    }
  }

  getPublicUrl(bucket: string, path: string): string {
    const { data } = supabase.storage
      .from(bucket)
      .getPublicUrl(path);

    return data.publicUrl;
  }

  validateFileSize(file: File, maxSizeMB: number): boolean {
    const maxSizeBytes = maxSizeMB * 1024 * 1024;
    return file.size <= maxSizeBytes;
  }

  validateFileType(file: File, allowedTypes: string[]): boolean {
    return allowedTypes.includes(file.type);
  }
}

export const storageService = new StorageService();

export const STORAGE_LIMITS = {
  avatars: {
    maxSize: 2,
    allowedTypes: ['image/jpeg', 'image/png', 'image/webp'],
  },
  portfolios: {
    maxSize: 5,
    allowedTypes: ['image/jpeg', 'image/png', 'image/webp'],
  },
  documents: {
    maxSize: 10,
    allowedTypes: ['application/pdf', 'image/jpeg', 'image/png'],
  },
  projectPhotos: {
    maxSize: 5,
    allowedTypes: ['image/jpeg', 'image/png', 'image/webp'],
  },
};

File: flutter_app/lib/services/storage_service.dart
import 'dart:io';
import 'package:supabase_flutter/supabase_flutter.dart';
import 'package:image_picker/image_picker.dart';
import 'package:file_picker/file_picker.dart';

class StorageService {
  final _supabase = Supabase.instance.client;
  final _imagePicker = ImagePicker();

  // Upload job photos
  Future<String?> uploadJobPhoto(String jobId, File file) async {
    try {
      final fileName = '${DateTime.now().millisecondsSinceEpoch}_${file.path.split('/').last}';
      final path = 'job-photos/$jobId/$fileName';

      await _supabase.storage.from('job-photos').upload(path, file);

      final url = _supabase.storage.from('job-photos').getPublicUrl(path);
      return url;
    } catch (e) {
      print('Error uploading job photo: $e');
      return null;
    }
  }

  // Upload multiple job photos
  Future<List<String>> uploadMultipleJobPhotos(String jobId, List<File> files) async {
    final urls = <String>[];
    for (final file in files) {
      final url = await uploadJobPhoto(jobId, file);
      if (url != null) urls.add(url);
    }
    return urls;
  }

  // Upload document (PDF, images)
  Future<String?> uploadDocument(String userId, File file) async {
    try {
      final fileName = '${DateTime.now().millisecondsSinceEpoch}_${file.path.split('/').last}';
      final path = 'documents/$userId/$fileName';

      await _supabase.storage.from('documents').upload(path, file);

      final url = _supabase.storage.from('documents').getPublicUrl(path);
      return url;
    } catch (e) {
      print('Error uploading document: $e');
      return null;
    }
  }

  // Upload portfolio image
  Future<String?> uploadPortfolioImage(String artisanId, File file) async {
    try {
      final fileName = '${DateTime.now().millisecondsSinceEpoch}_${file.path.split('/').last}';
      final path = 'portfolios/$artisanId/$fileName';

      await _supabase.storage.from('portfolios').upload(path, file);

      final url = _supabase.storage.from('portfolios').getPublicUrl(path);
      return url;
    } catch (e) {
      print('Error uploading portfolio image: $e');
      return null;
    }
  }

  // Delete file from storage
  Future<bool> deleteFile(String bucket, String path) async {
    try {
      await _supabase.storage.from(bucket).remove([path]);
      return true;
    } catch (e) {
      print('Error deleting file: $e');
      return false;
    }
  }

  // Pick image from gallery
  Future<File?> pickImageFromGallery() async {
    try {
      final XFile? image = await _imagePicker.pickImage(
        source: ImageSource.gallery,
        maxWidth: 1920,
        maxHeight: 1080,
        imageQuality: 85,
      );

      if (image != null) {
        return File(image.path);
      }
      return null;
    } catch (e) {
      print('Error picking image: $e');
      return null;
    }
  }

  // Pick multiple images
  Future<List<File>> pickMultipleImages({int maxImages = 5}) async {
    try {
      final List<XFile> images = await _imagePicker.pickMultiImage(
        maxWidth: 1920,
        maxHeight: 1080,
        imageQuality: 85,
      );

      final files = images.take(maxImages).map((xFile) => File(xFile.path)).toList();
      return files;
    } catch (e) {
      print('Error picking multiple images: $e');
      return [];
    }
  }

  // Pick image from camera
  Future<File?> takePhoto() async {
    try {
      final XFile? photo = await _imagePicker.pickImage(
        source: ImageSource.camera,
        maxWidth: 1920,
        maxHeight: 1080,
        imageQuality: 85,
      );

      if (photo != null) {
        return File(photo.path);
      }
      return null;
    } catch (e) {
      print('Error taking photo: $e');
      return null;
    }
  }

  // Pick document (PDF, etc)
  Future<File?> pickDocument() async {
    try {
      FilePickerResult? result = await FilePicker.platform.pickFiles(
        type: FileType.custom,
        allowedExtensions: ['pdf', 'jpg', 'jpeg', 'png'],
      );

      if (result != null && result.files.single.path != null) {
        return File(result.files.single.path!);
      }
      return null;
    } catch (e) {
      print('Error picking document: $e');
      return null;
    }
  }

  // Get file size in MB
  double getFileSizeInMB(File file) {
    final bytes = file.lengthSync();
    return bytes / (1024 * 1024);
  }

  // Validate file size (max 10MB)
  bool isFileSizeValid(File file, {double maxSizeMB = 10.0}) {
    return getFileSizeInMB(file) <= maxSizeMB;
  }
}
